### PR TITLE
Add support for EvalML

### PR DESCRIPTION
**Pull request by dsherry**
## Description

This PR seeks to add support for EvalML to BentoML.

## Motivation and Context

[EvalML](https://github.com/alteryx/evalml) is an automl library which unifies a variety of supervised ML libraries under one API. EvalML has its own core pipeline datastructure with an sklearn-style interface, including an API for defining execution graphs of transformers and estimators. EvalML's current transformer and estimator set includes components from sklearn, xgboost, lightgbm, catboost and more.

## How Has This Been Tested?

I've tried to mirror the testing pattern I see is currently in place in the BentoML repo for pytorch, spacy, onnx and other modeling packages.

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] New feature and improvements (non-breaking change which adds/improves functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Code Refactoring (internal change which is not user facing)
- [X] Documentation
- [X] Test, CI, or build

## Component(s) if applicable
- [ ] BentoService (service definition, dependency management, API input/output adapters)
- [X] Model Artifact (model serialization, multi-framework support)
- [ ] Model Server (mico-batching, dockerisation, logging, OpenAPI, instruments)
- [ ] YataiService gRPC server (model registry, cloud deployment automation)
- [ ] YataiService web server (nodejs HTTP server and web UI)
- [ ] Internal (BentoML's own configuration, logging, utility, exception handling)
- [ ] BentoML CLI

## Checklist:

- [x] My code follows the bentoml code style, both `./dev/format.sh` and
  `./dev/lint.sh` script have passed
  ([instructions](https://github.com/bentoml/BentoML/blob/master/DEVELOPMENT.md#style-check-and-auto-formatting-your-code)).
- [x] My change reduces project test coverage and requires unit tests to be added
- [X] I have added unit tests covering my code change
- [X] My change requires a change to the documentation
- [x] I have updated the documentation accordingly

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -139,4 +139,7 @@ __bentoml_tmp_sdist_build
 helm/YataiService/charts
 
 # Integration test data
-tests/integration/data
\ No newline at end of file
+tests/integration/data
+
+# Emacs buffers
+*~
```
Comment (dsherry) [inline .gitignore:145]: This change is unrelated to adding EvalML support. I am an emacs user and find this helpful.

File: bentoml/frameworks/sklearn.py [hunk 2]
```diff
@@ -33,7 +33,7 @@ class SklearnModelArtifact(BentoServiceArtifact):
         pickle_extension (str): The extension format for pickled file
 
     Raises:
-        MissingDependencyException: sklean package is required for SklearnModelArtifact
+        MissingDependencyException: sklearn package is required for SklearnModelArtifact
```
Comment (dsherry) [inline bentoml/frameworks/sklearn.py:36]: The minor docs changes in this file are unrelated to adding EvalML support. I noticed the typo and missing command and wanted to add a fix.

File: bentoml/frameworks/evalml.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+from bentoml.service.artifacts.common import PickleArtifact
+
+class EvalMLModelArtifact(BentoServiceArtifact):
+    """
+    Abstraction for saving/loading EvalML models
+
+    Args:
+        name (str): Name for the artifact
+        pickle_extension (str): The extension format for pickled file
+        pickle_module (str, module): pickle format to use: 'pickle' or 'cloudpickle'
```
Comment (dsherry) [inline bentoml/frameworks/evalml.py:15]: I saw that `PickleArtifact` supports both python native `pickle` and `cloudpickle`. So I wanted to give users the option here of using either format. This makes testing slightly more complex. I would gladly rip this out and stick to support for just one format if that is preferable to reviewers.

File: bentoml/frameworks/evalml.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        self._pickle_extension = pickle_extension
+        self._pickle_module = pickle_module
+        self._model = None
+
+    def _validate_package(self):
+        try:
+            from evalml import __version__ as evalml_version  # noqa # pylint: disable=unused-import
+        except ImportError:
+            raise MissingDependencyException(
+                "EvalML package is required to use EvalMLModelArtifact"
```
Comment (dsherry) [inline bentoml/frameworks/evalml.py:61]: Should this be raised at init-time or at runtime? I saw other artifact impls were raising at runtime so I added the same behavior here.

File: ci/evalml_integration_tests.sh [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+error=0
+trap 'error=1' ERR
+
+GIT_ROOT=$(git rev-parse --show-toplevel)
+cd "$GIT_ROOT" || exit
+
+# Install EvalML
+# TODO uncomment once evalml 0.24.0 is released, which adds support for pickle
+#pip install evalml>=0.24.0
+pip install git+https://github.com/alteryx/evalml.git@main
```
Comment (dsherry) [inline ci/evalml_integration_tests.sh:15]: TODO delete this change before getting reviews. This a testing hack until evalml v0.24.0 goes out, which adds support for native python `pickle` serialization.

File: tests/integration/test_evalml_model_artifact.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+import evalml
+
+import bentoml
+from bentoml.utils import cloudpickle
+from bentoml.yatai.client import YataiClient
+from tests.bento_service_examples.evalml_classifier import EvalMLClassifier, EvalMLClassifierCloudpickle
+
+
+@pytest.fixture(scope="session")
```
Comment (dsherry) [inline tests/integration/test_evalml_model_artifact.py:13]: I didn't see other integration tests doing this. But refactoring the other integration test model fixtures to use session scope (instead of the default which is function scope) could save on test runtime.

File: bentoml/frameworks/sklearn.py (hunk 2 above)
Comment (yubozhao) [reply bentoml/frameworks/sklearn.py:36]: Thank you!

File: bentoml/frameworks/evalml.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+    def load(self, path):
+        self._validate_package()
+        model_file_path = self._model_file_path(path)
+        try:
+            pickle_artifact = PickleArtifact(self.name,
+                                             pickle_module=self._pickle_module,
+                                             pickle_extension=self._pickle_extension)
+            pickle_artifact.load(path)
+            model = pickle_artifact.get()
```
Comment (yubozhao) [inline bentoml/frameworks/evalml.py:81]: Does EvalML provides its own save/load methods for model?

**Comment by yubozhao** [comment]
@dsherry I think this is a pretty completed PR.  I am wondering what's the advantage or difference between using this Model Artifact compare to Pickle Artifact.

File: bentoml/frameworks/evalml.py (hunk 7 above)
Comment (dsherry) [reply bentoml/frameworks/evalml.py:81]: @yubozhao yes, EvalML [defines save and load methods](https://github.com/alteryx/evalml/blob/main/evalml/pipelines/pipeline_base.py#L459) on its pipeline datastructure. We could certainly call that here. But I thought it would be convenient to use `PickleArtifact` instead.

Comment (yubozhao) [reply bentoml/frameworks/evalml.py:81]: If EvalML continue supported the API methods, I think it will be better to use the methods from the framework.

Comment (dsherry) [reply bentoml/frameworks/evalml.py:81]: Understood. I'll update this PR to use the EvalML `save`/`load` methods.

File: bentoml/frameworks/evalml.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+    def pack(self, evalml_model, metadata=None):  # pylint:disable=arguments-differ
+        self._validate_package()
+        self._model = evalml_model
+        return self
+
+    def load(self, path):
+        self._validate_package()
+        model_file_path = self._model_file_path(path)
+        if self._pipeline_base_class is None:
```
Comment (dsherry) [inline bentoml/frameworks/evalml.py:74]: Memoizing the import here felt like the safe thing to do. But if this is unnecessary please let me know and we can delete it.

**Comment by dsherry** [comment]
@yubozhao thanks for your comments. I've updated the implementation to use EvalML's [base pipeline `save` and `load`](https://evalml.alteryx.com/en/v0.24.0/generated/evalml.pipelines.PipelineBase.html#evalml.pipelines.PipelineBase) methods instead of simply using `PickleArtifact`. Let me know if there are any other changes we should make.

File: bentoml/frameworks/evalml.py (hunk 8 above)
Comment (yubozhao) [reply bentoml/frameworks/evalml.py:74]: I see what you are aiming at, that makes sense.

Since the `load` function should be only called once when BentoService is initializing it.  We can simplify this and just do inline import?

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -454,3 +454,20 @@ jobs:
         run: ./ci/pytorch_lightning_integration_tests.sh
       - name: Upload test coverage to Codecov
         uses: codecov/codecov-action@v1.0.12
+
+  onnx_integration_tests:
```
Comment (yubozhao) [inline .github/workflows/ci.yml:458]: Can you change the name to `eval_integration_tests`? Github CI requires unique name for each tasks

Comment (dsherry) [reply .github/workflows/ci.yml:458]: Oops, yes!

File: bentoml/frameworks/evalml.py (hunk 8 above)
Comment (dsherry) [reply bentoml/frameworks/evalml.py:74]: Ah, ok. Sure.

**Comment by yubozhao** [comment]
@dsherry Can you fix the linter issue and the permission issue for the EvalML test?  I think for the permission, you will want to make sure it have execute on it.  `chmod +x path/to/test.sh`

**Comment by yubozhao** [comment]
@dsherry Thank you for updating the PR. It looks great.  Can't wait the community try out this integration
